#411: Add asserts checking non-set commands: addresses #411

apoorvyadav1111 opened the pull request:
* The changes that are included in the PR.

Addresses #411 
Added assertType for commands that are not applicable for SET Data structures.

  * Design document, if any.
  
Commands related to STACKINT, Queue, BITs, GET were modified to return consistent messages.

  * Information on any implementation choices that were made.

  * Evidence of sufficient testing. You ``MUST`` indicate the tests done, either manually or automated.

existing test cases pass + added new test cases to check this behaviour.

lucifercr07 commented on config/main.go:
```diff
@@ -27,4 +27,3 @@ var (
 	// If requirepass is set to an empty string, no authentication is required
 	RequirePass string = constants.EmptyStr
 )
-
```
nit: Please fix

apoorvyadav1111 replied:
This was suggested by the linter after I ran the `golangci-lint run` command. Do I still remove this edit?

lucifercr07 replied:
Should be good then

lucifercr07 commented on core/eval.go:
```diff
@@ -261,6 +261,11 @@ func evalGET(args []string, store *Store) []byte {
 
+	// If the object exists, check if it is a set object.
+	if err := assertType(obj.TypeEncoding, ObjTypeSet); err == nil {
+		return diceerrors.NewErrSetType()
```
Please remove this function as encoding is to be done as part of `toEncodedMessage()`.
`WrongTypeErr` is already used at multiple places, we can follow the same pattern.
Ref: `return diceerrors.NewErrWithMessage(diceerrors.WrongTypeErr)`

apoorvyadav1111 replied:
Yes, there are multiple instances for this return, Would you prefer me to fix all of these returns and remove the NewErrSetType()?

lucifercr07 replied:
I think we should follow one pattern across the codebase. And implementation of `NewErrSetType` seems redundant in this case as we already have similar encoding method available and would make it harder from maintainability point also.
So yes we shall change it as `diceerrors.NewErrWithMessage(diceerrors.WrongTypeErr)` if possible.

lucifercr07 commented on core/eval.go:
```diff
@@ -261,6 +261,11 @@ func evalGET(args []string, store *Store) []byte {
 	}
 
+	// If the object exists, check if it is a set object.
+	if err := assertType(obj.TypeEncoding, ObjTypeSet); err == nil {
```
Shall L270 i.e `switch _, oEnc := ExtractTypeEncoding(obj); oEnc` shouldn't handle this case?
Any specific reasons to introduce this check here?

apoorvyadav1111 replied:
The existing switch statement does not return the message similar to the Redis implementation.

lucifercr07 replied:
In `default` it's having return error as `diceerrors.WrongTypeErr`.
I'd suggest having consistent way of having determining object encoding instead of multiple assert blocks.
@soumya-codes thoughts?

apoorvyadav1111 replied:
I think it checks the encoding and not the type of the value stored. GET currently expects only int and string as the value of the object (which is inline with redis implementation). In redis, the get functions only return strings. For now, the encoding of the SET values are similar to the ObjEncodingRaw's value, which is why it matches the second case. 

I am thinking following:
1. Create two new encodings for sets:

Current
```
var ObjEncodingHT uint8 = 0
```

New
```
var ObjEncodingSetInt uint8 = 11
var ObjEncodingSetStr uint8 = 12
```
2. Update the asserts for this new encoding

apoorvyadav1111 commented:
Hi @JyotinderSingh , @lucifercr07 

I have updated most of the commands to handle the error messages that did not return the wrong error type. I have updated the set data encoding (strings and ints, like Redis). 

Please review and share your views. If everything looks good, please merge.

Thanks,
Apoorv

lucifercr07 reviewed (approved):
LGTM

JyotinderSingh reviewed (approved):
Thanks for the improvements @apoorvyadav1111. Thanks for the reviews @lucifercr07.
